general/mem_manager.hpp

```diff
@@ -26,28 +26,64 @@ namespace mfem
 /// Memory types supported by MFEM.
 enum class MemoryType
 {
-   HOST,      ///< Host memory; using new[] and delete[]
-   HOST_32,   ///< Host memory aligned at 32 bytes (not supported yet)
-   HOST_64,   ///< Host memory aligned at 64 bytes (not supported yet)
-   CUDA,      ///< cudaMalloc, cudaFree
-   CUDA_UVM   ///< cudaMallocManaged, cudaFree (not supported yet)
+   HOST,           ///< Host memory; using new[] and delete[]
+   HOST_32,        ///< Host memory; aligned at 32 bytes
+   HOST_64,        ///< Host memory; aligned at 64 bytes
+   HOST_DEBUG,     ///< Host memory; protected while in device mode
+   HOST_UMPIRE,    ///< Host memory; using Umpire
+   HOST_MANAGED,   ///< Host memory; using *MallocManaged, *Free
+   DEVICE,         ///< Device memory; using *Malloc, *Free
+   DEVICE_DEBUG,   ///< Device memory; protected while in host mode
+   DEVICE_UMPIRE,  ///< Device memory; using Umpire
+   DEVICE_MANAGED, ///< Device memory; using D2D copies only
+   SIZE            ///< Number of host and device memory types
 };
 
-/// Memory classes identify subsets of memory types.
-/** This type is used by kernels that can work with multiple MemoryType%s. For
-    example, kernels that can use CUDA or CUDA_UVM memory types should use
-    MemoryClass::CUDA for their inputs. */
+/// Size of the memory types, the host and device ones.
+constexpr int MemoryTypeSize = static_cast<int>(MemoryType::SIZE);
+constexpr int HostMemoryTypeSize = static_cast<int>(MemoryType::DEVICE);
+constexpr int DeviceMemoryTypeSize = MemoryTypeSize - HostMemoryTypeSize;
+
+/// Memory type names, used during Device:: configuration.
+extern const char *MemoryTypeName[MemoryTypeSize];
+
+/// Pre & post increment operators for the MemoryType.
+inline MemoryType& operator++(MemoryType &mt)
+{ return mt = static_cast<MemoryType>(static_cast<int>(mt) + 1); }
+
+inline MemoryType& operator--(MemoryType &mt)
+{ return mt = static_cast<MemoryType>(static_cast<int>(mt) - 1); }
+
+inline MemoryType& operator++(MemoryType &mt,int)
+{ return mt = static_cast<MemoryType>(static_cast<int>(mt) + 1); }
+
+inline MemoryType& operator--(MemoryType &mt,int)
+{ return mt = static_cast<MemoryType>(static_cast<int>(mt) - 1); }
+
+
+/// Memory classes identify sets of memory types.
+/** This type is used by kernels that can work with multiple MemoryType%s.
+ *  For example, kernels that can use DEVICE or MANAGED memory types should
+ *  use MemoryClass::DEVICE for their inputs. */
 enum class MemoryClass
 {
-   HOST,    ///< Memory types: { HOST, HOST_32, HOST_64, CUDA_UVM }
+   HOST,    ///< Memory types: { HOST, HOST_32, HOST_64, MANAGED }
    HOST_32, ///< Memory types: { HOST_32, HOST_64 }
    HOST_64, ///< Memory types: { HOST_64 }
-   CUDA,    ///< Memory types: { CUDA, CUDA_UVM }
-   CUDA_UVM ///< Memory types: { CUDA_UVM }
+   DEVICE,  ///< Memory types: { DEVICE, MANAGED }
+   MANAGED  ///< Memory types: { MANAGED }
```

Comment:
The comments here need to be updated -- e.g. there is no `MANAGED` type at the moment, also the types `*_DEBUG` and `UMPIRE_*` are not included.

Reply:
`Updated`